Pass TinyGo wasi tests

TinyGo has a target `make test-tinygo-wasi` which ends up running the following via wasmtime. We should be able to do the same via wazero

```bash
$ tinygo test -target wasi -c compress/zlib container/heap container/list container/ring crypto/des crypto/internal/subtle crypto/md5 crypto/rc4 crypto/sha1 crypto/sha256 crypto/sha512 debug/macho embed/internal/embedtest encoding encoding/ascii85 encoding/base32 encoding/csv encoding/hex go/scanner hash hash/adler32 hash/crc64 hash/fnv html internal/itoa internal/profile math math/cmplx net net/http/internal/ascii net/mail os path reflect sync testing testing/iotest text/scanner unicode unicode/utf16 unicode/utf8  crypto/internal/nistec/fiat compress/bzip2 crypto/dsa index/suffixarray  ./tests/runtime_wasi
```

The only package tested that fails with wazero (so far) is "os"

```bash
./build/tinygo test -target wasi -c -o os.wasm $I; wazero run -mount=$(go env GOROOT)/src/os:/:ro -mount=/tmp:tmp -hostlogging=filesystem os.wasm
--snip--
==> wasi_snapshot_preview1.path_filestat_get(fd=3,flags=SYMLINK_FOLLOW,path=tmp/file)
<== (filestat={filetype=REGULAR_FILE,size=7,mtim=1673836571680643778},errno=ESUCCESS)
==> wasi_snapshot_preview1.path_filestat_get(fd=3,flags=,path=tmp/file)
<== (filestat={filetype=REGULAR_FILE,size=7,mtim=1673836571680643778},errno=ESUCCESS)
--- FAIL: TestChdir (0.00s)
    Mkdir(_os_test_TestChDir, 0755) returned mkdir _os_test_TestChDir: errno 52
    Chdir _os_test_TestChDir: chdir _os_test_TestChDir: errno 44
    FailNow is incomplete, requires runtime.Goexit()
FAIL

```

Full stack for TestChdir is

```
--> ._start()
	==> wasi_snapshot_preview1.environ_sizes_get(result.environc=172396,result.environv_len=172392)
	<== errno=ESUCCESS
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	--> .malloc(2)
	<-- 172672
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	--> .calloc(8,4)
		--> .malloc(32)
		<-- 172688
	<-- 172688
	--> .free(0)
	<--
	--> .malloc(1)
	<-- 172720
	--> .free(172672)
	<--
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
	==> wasi_snapshot_preview1.args_sizes_get(result.argc=172456,result.argv_len=172452)
	<== errno=ESUCCESS
	==> wasi_snapshot_preview1.args_get(argv=172736,argv_buf=172752)
	<== errno=ESUCCESS
	==> wasi_snapshot_preview1.clock_time_get(id=0,precision=1000,result.timestamp=172312)
	<== errno=ESUCCESS
	==> wasi_snapshot_preview1.clock_time_get(id=0,precision=1000,result.timestamp=170840)
	<== errno=ESUCCESS
	==> wasi_snapshot_preview1.clock_time_get(id=0,precision=1000,result.timestamp=170408)
	<== errno=ESUCCESS
	--> .realloc(0,20)
	<-- 179568
	--> .realloc(0,19)
	<-- 179600
	==> wasi_snapshot_preview1.path_unlink_file(fd=3,path=_os_test_TestChDir)
	<== errno=ENOENT
	==> wasi_snapshot_preview1.path_remove_directory(fd=3,path=_os_test_TestChDir)
	<== errno=ENOENT
	==> wasi_snapshot_preview1.path_create_directory(fd=3,path=_os_test_TestChDir)
	<== errno=EIO
	--> .realloc(0,19)
	<-- 180800
	==> wasi_snapshot_preview1.path_filestat_get(fd=3,flags=SYMLINK_FOLLOW,path=_os_test_TestChDir)
	<== (filestat=,errno=ENOENT)
	==> wasi_snapshot_preview1.path_unlink_file(fd=3,path=_os_test_TestChDir)
	<== errno=ENOENT
	==> wasi_snapshot_preview1.path_remove_directory(fd=3,path=_os_test_TestChDir)
	<== errno=ENOENT
	==> wasi_snapshot_preview1.clock_time_get(id=0,precision=1000,result.timestamp=170312)
	<== errno=ESUCCESS
	==> wasi_snapshot_preview1.clock_time_get(id=0,precision=1000,result.timestamp=170744)
	<== errno=ESUCCESS
--- FAIL: TestChdir (0.00s)
    Mkdir(_os_test_TestChDir, 0755) returned mkdir _os_test_TestChDir: errno 29
    Chdir _os_test_TestChDir: chdir _os_test_TestChDir: errno 44
    FailNow is incomplete, requires runtime.Goexit()
error instantiating wasm binary: module[] function[_start] failed: runtime error: index out of range [3] with length 0 (recovered by wazero)
wasm stack trace:
	wasi_snapshot_preview1.fd_write(i32,i32,i32,i32) i32
	.__wasi_fd_write(i32,i32,i32,i32) i32
	.write(i32,i32,i32) i32
	.(os.unixFileHandle).Write(i32,i32,i32,i32,i32)
		0xc72e: /Users/adrian/oss/tinygo/src/syscall/syscall_libc.go:32:16 (inlined)
		        /Users/adrian/oss/tinygo/src/os/exec_posix.go (inlined)
		        /Users/adrian/oss/tinygo/src/os/errors.go:25:18 (inlined)
		        /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/io/fs/fs.go:745:16 (inlined)
		        /Users/adrian/oss/tinygo/src/os/export_test.go:571:16 (inlined)
		        /Users/adrian/oss/tinygo/src/os/file_anyos.go:110:24 (inlined)
		        /Users/adrian/oss/tinygo/src/os/exec.go:569:20 (inlined)
		        /Users/adrian/oss/tinygo/src/os/stat_linux.go:52:7 (inlined)
		        /Users/adrian/oss/tinygo/src/os/types_anyos.go:48:23 (inlined)
		        /Users/adrian/oss/tinygo/src/os/types_anyos.go:46:21 (inlined)
		        /Users/adrian/oss/tinygo/src/os/stat_linux.go:59:7
	.(*os.File).Write(i32,i32,i32,i32,i32)
		0xd16b: /Users/adrian/oss/tinygo/src/os/file_anyos.go:109:25 (inlined)
		        <Go interface method> (inlined)
		        /Users/adrian/oss/tinygo/src/os/file.go:141:25 (inlined)
		        /Users/adrian/oss/tinygo/src/os/exec.go:569:20 (inlined)
		        /Users/adrian/oss/tinygo/src/os/stat_linux.go:52:7 (inlined)
		        /Users/adrian/oss/tinygo/src/os/types_anyos.go:48:23 (inlined)
		        /Users/adrian/oss/tinygo/src/os/types_anyos.go:46:21 (inlined)
		        /Users/adrian/oss/tinygo/src/os/stat_linux.go:59:7
	.interface:{Write:func:{slice:basic:uint8}{basic:int,named:error}}.Write$invoke(i32,i32,i32,i32,i32,i32)
		0xc07c: /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/fmt/print.go:187:13 (inlined)
		        /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/bufio/bufio.go (inlined)
		        <Go interface method>:25:18 (inlined)
		        <Go interface assert>:745:16 (inlined)
		        /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/bytes/buffer.go:571:16 (inlined)
		        /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/fmt/print.go:569:20 (inlined)
		        /Users/adrian/oss/tinygo/src/os/file.go:52:7 (inlined)
		        /Users/adrian/oss/tinygo/src/reflect/value.go:48:23 (inlined)
		        /Users/adrian/oss/tinygo/src/reflect/value.go:46:21 (inlined)
		        /Users/adrian/oss/tinygo/src/os/file.go:59:7
	.(*bytes.Buffer).WriteTo(i32,i32,i32,i32)
		0x36f19: /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/bytes/buffer.go:252:18 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/slice.go (inlined)
		         /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/bytes/bytes.go:25:18 (inlined)
		         /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/bytes/buffer.go:745:16 (inlined)
	.testing.tRunner(i32,i32,i32)
		0x50875: /Users/adrian/oss/tinygo/src/testing/testing.go:94:19 (inlined)
		         /Users/adrian/oss/tinygo/src/os/file_anyos.go (inlined)
		         /Users/adrian/oss/tinygo/src/testing/fuzz.go:25:18 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/benchmark.go:745:16 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/testing.go (inlined)
		         /Users/adrian/oss/tinygo/src/testing/testing.go:479:13 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/testing.go:370:10 (inlined)
		         /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/io/fs/fs.go:234:37 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/testing.go:548:17 (inlined)
		         /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/time/format.go:23:23 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/testing.go:141:25 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/testing.go:120:26 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/testing.go:101:24 (inlined)
	.(*testing.M).Run(i32) i32
		0x23154: /Users/adrian/oss/tinygo/src/testing/testing.go:466:9 (inlined)
		         /Users/adrian/oss/tinygo/src/os/file_anyos.go (inlined)
		         /Users/adrian/oss/tinygo/src/testing/fuzz.go:25:18 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/benchmark.go:745:16 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/testing.go:444:29 (inlined)
		         /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/math/unsafe.go:571:16 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/testing.go:141:25 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/match.go:569:20 (inlined)
		         /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/strconv/atoi.go:52:7 (inlined)
		         <Go interface assert>:48:23 (inlined)
		         <Go interface assert>:46:21 (inlined)
		         /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/strconv/atoi.go:59:7 (inlined)
		         /Users/adrian/oss/tinygo/src/testing/match.go:47:19 (inlined)
		         /Users/adrian/.gimme/versions/go1.19.5.darwin.amd64/src/math/abs.go
	.runtime.run$1()
		0x1f1ac: /Users/adrian/Library/Caches/go-build/40/40ab9d36559948f2c35801c3c43c29352ce15c43d8e4770a86b7ca1dbf966aa5-d:115:15 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/panic.go:59:7 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/arch_tinygowasm_malloc.go:47:19 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/scheduler.go
	.runtime.run$1$gowrapper(i32)
		0x1e01b: /Users/adrian/oss/tinygo/src/runtime/scheduler_any.go:23:2 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/panic.go:59:7 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/arch_tinygowasm_malloc.go:47:19 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/scheduler.go
	.tinygo_launch(i32)
	._start()
		0x1dec9: /Users/adrian/oss/tinygo/src/internal/task/task_asyncify.go:109:17 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/panic.go:59:7 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/arch_tinygowasm_malloc.go:47:19 (inlined)
		         /Users/adrian/oss/tinygo/src/runtime/scheduler.go
```

I narrowed it down to something odd about a function return not having any of the parameters at the time it is returning. There's an index out of bounds when logging is on, though the test also fails with it off.

[os.wasm.gz](https://github.com/tetratelabs/wazero/files/10422242/os.wasm.gz)

wazero built with the below patch
```bash
$ wazero run -mount=/usr/local/go/src/os:/ /Users/adrian/oss/tinygo/os.wasm -test.run '^TestChdir$' 
--- FAIL: TestChdir (0.00s)
    Mkdir(_os_test_TestChDir, 0755) returned mkdir _os_test_TestChDir: errno 29
    Chdir _os_test_TestChDir: chdir _os_test_TestChDir: errno 44
    FailNow is incomplete, requires runtime.Goexit()
FAIL
$ wazero run -hostlogging=filesystem -mount=/usr/local/go/src/os:/ /Users/adrian/oss/tinygo/os.wasm -test.run '^TestChdir$' 
==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
<== (prestat={pr_name_len=1},errno=ESUCCESS)
==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
<== (path=/,errno=ESUCCESS)
==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
<== (prestat=,errno=EBADF)
==> wasi_snapshot_preview1.path_unlink_file(fd=3,path=_os_test_TestChDir)
<== errno=ENOENT
==> wasi_snapshot_preview1.path_remove_directory(fd=3,path=_os_test_TestChDir)
<== errno=ENOENT
==> wasi_snapshot_preview1.path_create_directory(fd=3,path=_os_test_TestChDir)
<== errno=EIO
==> wasi_snapshot_preview1.path_filestat_get(fd=3,flags=SYMLINK_FOLLOW,path=_os_test_TestChDir)
<== (filestat=,errno=ENOENT)
==> wasi_snapshot_preview1.path_unlink_file(fd=3,path=_os_test_TestChDir)
<== errno=ENOENT
==> wasi_snapshot_preview1.path_remove_directory(fd=3,path=_os_test_TestChDir)
<== errno=ENOENT
--- FAIL: TestChdir (0.00s)
    Mkdir(_os_test_TestChDir, 0755) returned mkdir _os_test_TestChDir: errno 29
    Chdir _os_test_TestChDir: chdir _os_test_TestChDir: errno 44
    FailNow is incomplete, requires runtime.Goexit()
FAIL
```

patch about the index out of bounds
```diff
diff --git a/cmd/wazero/wazero.go b/cmd/wazero/wazero.go
index e22cf3c1..f7a375e9 100644
--- a/cmd/wazero/wazero.go
+++ b/cmd/wazero/wazero.go
@@ -306,7 +306,7 @@ func maybeHostLogging(ctx context.Context, hostLogging *string, stdErr logging.W
        switch h {
        case "":
        case "filesystem":
-               ctx = context.WithValue(ctx, experimental.FunctionListenerFactoryKey{}, logging.NewFilesystemLoggingListenerFactory(stdErr))
+               ctx = context.WithValue(ctx, experimental.FunctionListenerFactoryKey{}, logging.NewHostLoggingListenerFactory(stdErr))
        default:
                fmt.Fprintf(stdErr, "invalid hostLogging value: %v\n", h)
                exit(1)
diff --git a/internal/wasi_snapshot_preview1/logging/logging.go b/internal/wasi_snapshot_preview1/logging/logging.go
index de5d1372..4ca2d3d2 100644
--- a/internal/wasi_snapshot_preview1/logging/logging.go
+++ b/internal/wasi_snapshot_preview1/logging/logging.go
@@ -123,6 +123,10 @@ func logErrno(_ context.Context, _ api.Module, w logging.Writer, _, results []ui
 type logMemI32 uint32
 
 func (i logMemI32) Log(_ context.Context, mod api.Module, w logging.Writer, params []uint64) {
+       if len(params) <= int(i) {
+               println(i)
+               return
+       }
        if v, ok := mod.Memory().ReadUint32Le(uint32(params[i])); ok {
                writeI32(w, v)
        }
diff --git a/internal/wasmdebug/debug.go b/internal/wasmdebug/debug.go
index 285a7717..6bc9ba8d 100644
--- a/internal/wasmdebug/debug.go
+++ b/internal/wasmdebug/debug.go
@@ -113,7 +113,7 @@ type stackTrace struct {
 }
 
 func (s *stackTrace) FromRecovered(recovered interface{}) error {
-       if false {
+       if true {
                debug.PrintStack()
        }
```

--- a/.github/workflows/integration.yaml
+++ b/.github/workflows/integration.yaml
@@ -116,7 +116,7 @@ jobs:
 
       - name: Run standard library tests
         # On PRs, we only test the minimal std packages to test basic functionality.
-        run: tinygo test -target ./wasi_wazero.json os net text/template
+        run: tinygo test -target ./wasi_wazero.json os net math
 
       - name: Run all standard library tests
         # Running all tests is pretty slow (~8min), so only do it on the main branch.

--- a/.github/workflows/integration.yaml
+++ b/.github/workflows/integration.yaml
@@ -44,7 +44,7 @@ jobs:
           path: |
             zig-cache
             zig-out
-          key: ${{ env.ZIG_BUILD_VERSION }}
+          key: integration-test-zig-${{ env.ZIG_BUILD_VERSION }}
 
       - name: Build Stdlib test binary
         # TODO: maybe uploading the built binary to GHA artifact would be helpful to reproduce the failure locally.
@@ -70,3 +70,101 @@ jobs:
           go run ./cmd/wazero run -hostlogging=filesystem -mount=.:/ ${{ steps.zig-build-test-binary.outputs.ZIG_BINARY_PATH }}
         # TODO: remove when #1029 is complete
         continue-on-error: true
+
+  tinygo:
+    name: TinyGo
+    runs-on: ubuntu-20.04
+    env:
+      # Until https://github.com/tinygo-org/tinygo/pull/3399 is released in 0.27.0,
+      # use the artifact from the GitHub Actions on the TinyGo repository.
+      # 3961571741 matches the one of the merge commit for the PR ^.
+      # https://github.com/tinygo-org/tinygo/actions/runs/3961571741
+      TINYGO_RUN_ID: 3961571741
+
+    steps:
+      - uses: actions/cache@v3
+        id: cache
+        with:
+          path:
+            ~/.cache/go-build
+            ~/.cache/tinygo
+          key: integration-test-tinygo-${{ env.TINYGO_RUN_ID }}-${{ env.GO_VERSION }}
+
+      - uses: actions/setup-go@v3
+        with:
+          go-version: ${{ env.GO_VERSION }}
+
+      - name: Install TinyGo
+        if: steps.cache.outputs.cache-hit != 'true'
+        env:
+          # Necessary to use gh command.
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          gh run download --repo tinygo-org/tinygo ${{ env.TINYGO_RUN_ID }} -n linux-amd64-double-zipped
+          tar -xzf tinygo.linux-amd64.tar.gz
+          sudo dpkg -i tinygo_amd64.deb
+
+      - name: Checkout wazero
+        uses: actions/checkout@v3
+
+      - name: Install wazero
+        run: go install ./cmd/wazero
+
+      - name: Generate target json for wazero.
+        run: |
+          jq '.emulator = "wazero run -hostlogging=filesystem -mount=.:/ -mount={tmpDir}:/tmp {}"' \
+            $(tinygo env TINYGOROOT)/targets/wasi.json > ./wasi_wazero.json
+
+      - name: Run standard library tests
+        # On PRs, we only test the minimal std packages to test basic functionality.
+        run: tinygo test -target ./wasi_wazero.json os math hash
+
+      - name: Run all standard library tests
+        # Running all tests is pretty slow (~8min), so only do it on the main branch.
+        if: ${{ github.event_name  == 'push' }}
+        # The list of packages are derived from:
+        # https://github.com/tinygo-org/tinygo/blob/v0.26.0/Makefile#L271-L319
+        # Note: index/suffixarray is extremely slow, so skip it.
+        run: |
+          tinygo test -target ./wasi_wazero.json \
+            compress/zlib \
+            container/heap \
+            container/list \
+            container/ring \
+            crypto/des \
+            crypto/internal/subtle \
+            crypto/md5 \
+            crypto/rc4 \
+            crypto/sha1 \
+            crypto/sha256 \
+            crypto/sha512 \
+            debug/macho \
+            embed/internal/embedtest \
+            encoding \
+            encoding/ascii85 \
+            encoding/base32 \
+            encoding/csv \
+            encoding/hex \
+            go/scanner \
+            hash \
+            hash/adler32 \
+            hash/crc64 \
+            hash/fnv \
+            html \
+            internal/itoa \
+            internal/profile \
+            math \
+            math/cmplx \
+            net \
+            net/http/internal/ascii \
+            net/mail \
+            os \
+            path \
+            reflect \
+            sync \
+            testing \
+            testing/iotest \
+            text/scanner \
+            unicode \
+            unicode/utf16 \
+            unicode/utf8

--- a/.github/workflows/integration.yaml
+++ b/.github/workflows/integration.yaml
@@ -111,12 +111,12 @@ jobs:
 
       - name: Generate target json for wazero.
         run: |
-          jq '.emulator = "wazero run -hostlogging=filesystem -mount=.:/ -mount={tmpDir}:/tmp {}"' \
+          jq '.emulator = "wazero run -mount=.:/ -mount={tmpDir}:/tmp {}"' \
             $(tinygo env TINYGOROOT)/targets/wasi.json > ./wasi_wazero.json
 
       - name: Run standard library tests
         # On PRs, we only test the minimal std packages to test basic functionality.
-        run: tinygo test -target ./wasi_wazero.json os math hash
+        run: tinygo test -target ./wasi_wazero.json os net text/template
 
       - name: Run all standard library tests
         # Running all tests is pretty slow (~8min), so only do it on the main branch.

--- a/.github/workflows/integration.yaml
+++ b/.github/workflows/integration.yaml
@@ -95,7 +95,6 @@ jobs:
           go-version: ${{ env.GO_VERSION }}
 
       - name: Install TinyGo
-        if: steps.cache.outputs.cache-hit != 'true'
         env:
           # Necessary to use gh command.
           GH_TOKEN: ${{ github.token }}

